### DISCUSSION TITLE
Serverless v4 at least 10x my lambda size

### DISCUSSION THREAD

**Discussion by kyusungpark** [serverless/serverless#12946] [Q&A]
Hey everyone,

I just noticed that my all lambdas had their zip size increase at least by 10x than what was before on v3. Has anyone ever experience this issue after switching to V4? On v3, I had serverless-esbuild and everything was good without any special configurations. V4 has serverless-esbuild built in and i was expecting it to work the same, but it obviosly is not.

Tried using same config [here](https://www.serverless.com/framework/docs/providers/aws/guide/building), but no change in lambda zip size.

Any recommendations for change?

My prev config. Added no additional config for serverless-esbuild
```
  package: {
    individually: true, // bundle size affects cold starts: package individually
    patterns: ['!package-lock.json'],
    exclude: ['**/node_modules/**'],
  },
```

Tried to add default config with v4 to see if there any changes but no changes in lambda size.
```
  build: {
    esbuild: {
      bundle: true,
      minify: true,
      external: ['@aws-sdk/client-s3'],
      exclude: ['@aws-sdk/*'],
      packages: 'external',
      buildConcurrency: 3,
      sourcemap: {
        type: 'linked',
        setNodeOptions: true,
      },
    },
  },
```

**Comment by kyusungpark** [comment]
I don't think it's treeshaking properly and it's including libraries that aren't necessary for each lambda. How do you properly tree-shake with serverless v4?

**Reply by kyusungpark** [reply]
For anyone having the same problem, turn off default esbuild that comes with serverless V4 and use serverless-esbuild. Seems like default esbuild that comew with v4 is unstable

**Comment by MukundaKatta** [comment]
v4's built-in esbuild defaults aren't as aggressive about treeshaking as the v3 plugin's defaults were. Add `build.esbuild` in your `serverless.yml` with `minify: true`, `treeShake: true`, and `external` for anything pulled from Lambda layers — most 10x regressions I've seen are one of those three missing. If you want the v3-identical behaviour, you can still install `serverless-esbuild` as a plugin and it overrides the built-in.

**Comment by tamish-max** [comment]
A 10x jump usually means the new build is no longer bundling the same way your v3 setup was.

The main red flag in the config you posted is:

```yml
packages: external
```

That tells esbuild not to bundle dependencies, which often leads to much fatter zip files because the packaging step carries far more of `node_modules`.

I’d try the opposite first and only externalize the few packages you intentionally want outside the bundle:

```yml
package:
  individually: true

build:
  esbuild:
    bundle: true
    minify: true
    sourcemap: false
    exclude:
      - '@aws-sdk/*'
    # remove packages: external
```

A quick way to confirm the culprit is:

1. run `serverless package`
2. unzip one artifact in `.serverless/`
3. compare its contents to your old v3 artifact
4. look for full packages that used to be tree-shaken/bundled

If you depend on native modules or Lambda layers, mark only those as external. Otherwise, letting esbuild fully bundle each function usually produces the smaller zip.
